runtime/vm/xz64/UpcallThunkGen.cpp

```diff
@@ -0,0 +1,922 @@
+ * Format: BRCL M1, RI2
+ * M1 - conditionMask
+ * RI2 - immediateInteger
+ */
+static I_32
+BRCL(I_8 *instructionPtr, I_32 conditionMask, I_32 immediateInteger)
+{
+	I_32 instructionSize = 6;
+	I_32 opcode = 0xC0040000;
+	*((I_32 *)instructionPtr) = (opcode | ((conditionMask & 0xF) << 20) | ((immediateInteger & 0xFFFF0000) >> 16));
+	instructionPtr[4] = (immediateInteger & 0xFF00) >> 8;
+	instructionPtr[5] = immediateInteger & 0xFF;
+	return instructionSize;
+}
+
+/*
+ * Generates an instruction to add a 32-bit integer constant to the 32-bit value in R1
+ * Format: "AFI R1,I2"
+ * R1 - sourceRegister
+ * I2 - integerConstant
+ */
+static I_32
+AFI(I_8 *instructionPtr, I_32 sourceRegister, I_32 immediateInteger)
+{
+	I_32 instructionSize = 6;
+	I_32 opcode = 0xC2090000;
+	*((I_32 *)instructionPtr) = (opcode | ((sourceRegister & 0x0000000F) << 20) | ((immediateInteger & 0xFFFF0000) >> 16));
+	instructionPtr[4] = (I_8)((immediateInteger & 0xFF00) >> 8);
+	instructionPtr[5] = (I_8)(immediateInteger & 0xFF);
+	return instructionSize;
+
+}
+
+/*
+ * Generates a sequence of instructions to copy data from a Java MemorySegment Object to a C struct.
+ * The struct is smaller thani or equal to 4096 bytes.
+ */
+static I_32
+generateSmallStructCopyInstructions(I_8 *instructionPtr, I_32 structSize, I_32 offsetToParameterArea)
+{
+	I_32 totalInstructionSize = 0;
+        // Load R4 with the hiddenParameter address
+        totalInstructionSize += LG(instructionPtr, R4, 0, R15, offsetToParameterArea);
+
+	const I_32 numLargeCopies = structSize / maxMVCCopyLength;
+        const I_32 residue = structSize % maxMVCCopyLength;
+
+	for(int i = 0; i < numLargeCopies; i++) {
+		// MVC D1(L,B1),D2(B2)
+		// The 'L' field is 0-based
+		totalInstructionSize += MVC(instructionPtr + totalInstructionSize, i*maxMVCCopyLength, maxMVCCopyLength - 1, R4, i*maxMVCCopyLength, R2);
+	}
+	//printf("generateStructCopyInstructions_0: numLargeCopies: %d, residue: %d, structSize: %d, offsetParameter: %d\n", numLargeCopies, residue, structSize, offsetToParameterArea);
+	if (residue != 0) {
+		// MVC D1(L,B1),D2(B2)
+		totalInstructionSize += MVC(instructionPtr + totalInstructionSize, numLargeCopies * maxMVCCopyLength, residue - 1, R4, numLargeCopies * maxMVCCopyLength, R2);
+	}
+	return totalInstructionSize;
+}
+
+/*
+ * Generates a sequence of instructions to copy data from a Java MemorySegment Object to a C struct.
+ * The struct is larger than 4096 bytes.
+ */
+static I_32
+generateLargeStructCopyInstructions(I_8 *instructionPtr, I_32 structSize, I_32 offsetToParameterArea)
+{
+	I_32 totalInstructionSize = 0;
+
+	const I_32 numFullSlots = structSize / maxMVCCopyLength;
+        const I_32 residue = structSize % maxMVCCopyLength;
+
+	const I_32 maxLargeCopies = 16;
+	const I_32 loopInstructionsLength = 114; // size in bytes
+	const I_32 sourceRegister = R2;
+	const I_32 destinationRegister = R4;
```

Comment:
This means the function is fixed to use `R2` as source register and `R4` as destination register deeming it to be used in specific locations only. Should this be parameterized to allow it to be used if we ever have to copy with different source and destination register?

Reply:
Now the source and destination register operands are parametrized for the struct copy routines: 109cbe6452e154afe9686653d3693605391b8423